### PR TITLE
When squashing commits in web UI, only compare emails

### PR DESCRIPTION
**Pull request by n0toose**
When squashing commits in web UI, only compare emails
    
Say that a user uses the email address "user@example.com" in their
local Git configuration, as well as their Gitea profile. Right now,
if they use a different name for their profile than the one that
they use in their local Git configuration (e.g. they make commits
using their full name, but their profile contains a different
for a display name pseudonym, then the UI will act as if there
are two separate identities, even if they use the same email address.
    
This doesn't fix the underlying problem completely (user accounts
can contain multiple email addresses), but it improves it.

### CODE REVIEW

**Comment by n0toose** [comment]
How I tested this:

- Used my ordinary name locally, but used a funny display name on Gitea like "HI IM GITEA". No `Co-authored-by` prompt appeared.
- `git commit --amend --author="Hello World <hello@world.com>", force-pushed, attempted to create a new PR again, I did see the `Co-authored-by:` (but not `Co-commited-by:`) trailer when I attempted to squash, so I'm assuming that my change works.

**Comment by n0toose** [comment]
Additional case scenario that I didn't consider:

User has two emails:
- `user1@example.com` (primary)
- `user1+jrwqijrwoq@example.com` (backup)

User uses `user1+jrwqijrwoq@example.com` in Git commit. Gitea can't recognize that the person with the email address `user1+jrwqijrwoq@example.com` is the same one as the one who created the PR.

**Comment by n0toose** [comment]
Problem in action:
- Commit: https://codeberg.org/Codeberg/Documentation/commit/e91d2ea42e68c191202a4a1007377f7586e38dc7
- Pull Request: https://codeberg.org/Codeberg/Documentation/pulls/306

Keep in mind that I got super confused over there for a sec, I ended up remembering about this quirk of Gitea only after merging the PR despite having noticed it at least once before.

**Comment by silverwind** [comment]
Can we compare these emails [case-insensitively](https://pkg.go.dev/strings#EqualFold)?

**Comment by n0toose** [comment]
I actually spoke about this with a friend as I was authoring the patch you instructed me to write, and he reassured me that the host part of an email address is case-insensitive, whereas the local part of an email address is case-sensitive. I checked: https://www.rfc-editor.org/rfc/rfc1035

**Comment by n0toose** [comment]
Looking at how I could make this compliant.

**Comment by n0toose** [comment]
Tested the "Hello World" author commit. I made this push very hesitantly as I've never worked with these Go functions before (but hey, maybe the fact that I found a way to do this so fast means that I'm learning!) and I am not sure whether this is actually working properly or the way I think it is.

**Comment by n0toose** [comment]
Okay, I tested it with a commit that uses the email of the Gitea account as the author, and it seems to work, just like it did earlier:

image

**Comment by silverwind** [comment]
> I actually spoke about this with a friend as I was authoring the patch you instructed me to write, and he reassured me that the host part of an email address is case-insensitive, whereas the local part of an email address is case-sensitive

All parts of an e-mail adress are [practically case insensitive](https://stackoverflow.com/a/9808332/808699), at least that's how 99% of mail providers implement it. [rfc1035](https://www.rfc-editor.org/rfc/rfc1035) is probably out of date in this regard.

**Comment by n0toose** [comment]
True. Okay, let's do this another way, does Gitea assume that email addresses are unconditionally 100% case-insensitive?

File: services/pull/merge_squash.go [hunk 1]
```diff
@@ -63,8 +64,9 @@ func doMergeStyleSquash(ctx *mergeContext, message string) error {
 		return err
 	}
 
-	if setting.Repository.PullRequest.AddCoCommitterTrailers && ctx.committer.String() != sig.String() {
-		// add trailer
+	if setting.Repository.PullRequest.AddCoCommitterTrailers &&
+		strings.EqualFold(ctx.committer.Email, sig.Email) {
+		// add trailer if email of user and email of committer don't match
```
Comment (wxiaoguang) [inline services/pull/merge_squash.go:69]: `don't match`, do you mean `!` `strings.EqualFold` ?

Comment (n0toose) [reply services/pull/merge_squash.go:69]: You're right. If I was under the impression that this still worked even if it didn't when I tested it, I seriously need to re-examine what I just did.

**Comment by n0toose** [comment]
Closing for now to save resources on your end, I will reopen this PR and push further commits if I have something else to add.

**Pull request closed without merging**